Runtime metrics docs (#326)
signalfx/signalfx-dotnet-tracing: docs/internal/runtime-metrics.md

# About the .NET runtime metrics

> The SignalFx Instrumentation for .NET includes an automatic runtime metrics collection.

Reviewer: ```suggestion
The SignalFx Instrumentation for .NET is able to collect runtime metrics.
```

> This feature can be enabled with configuration setting and is disabled by default. 

Reviewer: ```suggestion
This feature can be enabled using configuration settings.
```

Reviewer: Can we configure the interval?

Author: its fixed

> When enabled, metrics are periodically captured and sent to Splunk Observability Cloud.

Reviewer: ```suggestion
Metrics are exported on a periodic basis (push model).
```

Reviewer: I propose the change as they can be sent to the collector.
maybe

```suggestion
When enabled, metrics are periodically captured and sent to the configured endpoint.
```

Reviewer 2: +1 Although from a customer perspective, the end goal is getting them to O11y Cloud, either via Collector or directly.

Author: leaving it as it is for now

# Enable runtime metrics

To enable runtime metrics, set the `SIGNALFX_RUNTIME_METRICS_ENABLED` environment variable
to `true` for your .NET process. 

# Metrics exporter settings

## Default configuration
Metrics are sent to local [Splunk OpenTelemetry Collector](https://github.com/signalfx/splunk-otel-collector)
that forwards them to Splunk Observability Cloud. For guidance on how to setup required components of metrics pipeline in the collector, see 
[receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/signalfxreceiver) and [exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/signalfxexporter) documentation.

> ## Exporting directly to Splunk Observabilty Cloud

Reviewer: there should be a blank line after the header (applies to other places)

> If you prefer to send metrics directly to Splunk Observability Cloud, configure the following settings:

Reviewer: I find it hard to process that two possibilities are described at the same time. @theletterf WDYT?

Reviewer 2: You mean Collector + Direct ingest? We may want to clarify when direct ingest should be used first (for example, when installing a collector is not possible). How would you prefer this?

Reviewer: I mean that it describes that direct ingest can be configured by specifying
- SIGNALFX_ACCESS_TOKEN and SIGNALFX_REALM
OR
- SIGNALFX_ACCESS_TOKEN and SIGNALFX_METRICS_ENDPOINT_URL

Reviewer 2: Made an edit in the IMPORTANT note below for that.

| Setting | Value | Notes |
|-|-|-|
| `SIGNALFX_ACCESS_TOKEN` | *organization access tokens* | (__required__) See [here](https://docs.splunk.com/Observability/admin/authentication-tokens/org-tokens.html) to learn how to obtain one. |

> | `SIGNALFX_REALM` | *ingest realm* | (__recommended__) Set to send telemetry directly to Splunk Observability Cloud. If configured, metrics are sent to `https://ingest.<SIGNALFX_REALM>.signalfx.com/v2/datapoint`. To find your realm, open Splunk Observability Cloud, click Settings, and click on your user name. __NOTE: affects traces endpoint as well.__ |

Reviewer 3: as I remeber @theletterf preferred different wording for ingest.
@theletterf, could you please check it?

| `SIGNALFX_METRICS_ENDPOINT_URL` | *valid endpoint url* | (__optional__) Overrides other configuration settings for metrics ingestion endpoint. |

__IMPORTANT:__ One of `SIGNALFX_REALM` or `SIGNALFX_METRICS_ENDPOINT_URL` has to be configured in addition to `SIGNALFX_ACCESS_TOKEN`.

# Troubleshooting

## How to check if runtime metrics are enabled?

The SignalFx Instrumentation for .NET logs the profiling configuration at `INF` during the startup. 
Verify that `runtime_metrics_enabled` key is set to `true` under `TRACER CONFIGURATION`.

## How to check where are metrics exported?

Check the value of the `metrics_agent_url` key under `TRACER CONFIGURATION`.

## How to verify collector setup?

* Make sure you are running the [Splunk OpenTelemetry Collector](https://github.com/signalfx/splunk-otel-collector)
* Make sure that the collector is configured correctly to handle metrics.

> * Make sure that an `signalfx` receiver and `signalfx` exporter are configured in the collector.

Reviewer: is it not a description of the line above?

```suggestion
Make sure that a `signalfx` receiver and a `signalfx` exporter are configured in the collector.
```

or maybe the previous line should be removed?

Author: previous line removed

# .NET runtime metrics

The SignalFx Instrumentation for .NET includes automatic runtime metrics collection.
This feature can be enabled using configuration settings.
When enabled, metrics are periodically captured and sent to Splunk Observability Cloud.

## Enable runtime metrics

To enable runtime metrics, set the `SIGNALFX_RUNTIME_METRICS_ENABLED` environment variable
to `true` for your .NET process.

## Metrics exporter settings

### Default configuration

Metrics are sent to the local instance of [Splunk OpenTelemetry Collector](https://github.com/signalfx/splunk-otel-collector),
which forwards telemetry to Splunk Observability Cloud. For instructions on how to set up required components of the metrics pipeline in the collector, see 
[receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/signalfxreceiver) and [exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/signalfxexporter) documentation.

### Exporting directly to Splunk Observability Cloud

> If you prefer to send metrics directly to Splunk Observability Cloud, configure the following settings:

Reviewer 2: ```suggestion
If you prefer to send metrics to Splunk Observability Cloud without using the Collector, configure the following settings:
```

## About the .NET runtime metrics

The SignalFx Instrumentation for .NET includes an automatic runtime metrics collection.
This feature can be enabled with configuration setting and is disabled by default. 
When enabled, metrics are periodically captured and sent to Splunk Observability Cloud.

## Enable runtime metrics

To enable runtime metrics, set the `SIGNALFX_RUNTIME_METRICS_ENABLED` environment variable
to `true` for your .NET process.

## Metrics exporter settings

### Default configuration

Metrics are sent to local [Splunk OpenTelemetry Collector](https://github.com/signalfx/splunk-otel-collector)
that forwards them to Splunk Observability Cloud. For guidance on how to setup required components of metrics pipeline in the collector, see 
[receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/signalfxreceiver) and [exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/signalfxexporter) documentation.

### Exporting directly to Splunk Observabilty Cloud

If you prefer to send metrics directly to Splunk Observability Cloud, configure the following settings:

| Setting | Value | Notes |
|-|-|-|

> | `SIGNALFX_ACCESS_TOKEN` | *organization access tokens* | (__required__) See [here](https://docs.splunk.com/Observability/admin/authentication-tokens/org-tokens.html) to learn how to obtain one. |

Reviewer 2: ```suggestion
| `SIGNALFX_ACCESS_TOKEN` | Your Observability Cloud token | (__required__) See [here](https://docs.splunk.com/Observability/admin/authentication-tokens/org-tokens.html) to learn how to obtain one. |
```

> | `SIGNALFX_REALM` | *ingest realm* | (__recommended__) Set to send telemetry directly to Splunk Observability Cloud. If configured, metrics are sent to `https://ingest.<SIGNALFX_REALM>.signalfx.com/v2/datapoint`. To find your realm, open Splunk Observability Cloud, click Settings, and click on your user name. __NOTE: affects traces endpoint as well.__ |

Reviewer 2: ```suggestion
| `SIGNALFX_REALM` | Your Observability Cloud realm | Set to send telemetry directly to Splunk Observability Cloud. If configured, metrics are sent to `https://ingest.<SIGNALFX_REALM>.signalfx.com/v2/datapoint`. To find your realm, open Splunk Observability Cloud, click Settings, then click on your username. Setting this variable also modify the default tracing endpoint. |
```

> | `SIGNALFX_METRICS_ENDPOINT_URL` | *valid endpoint url* | (__optional__) Overrides other configuration settings for metrics ingestion endpoint. |

Reviewer 2: ```suggestion
| `SIGNALFX_METRICS_ENDPOINT_URL` | Metric endpoint | (Optional) Overrides other settings for the metrics ingestion endpoint. |
```

> __IMPORTANT:__ One of `SIGNALFX_REALM` or `SIGNALFX_METRICS_ENDPOINT_URL` has to be configured in addition to `SIGNALFX_ACCESS_TOKEN`.

Reviewer 2: ```suggestion
__IMPORTANT:__ To export data directly to Splunk Observability Cloud, set `SIGNALFX_REALM` or `SIGNALFX_METRICS_ENDPOINT_URL` in addition to `SIGNALFX_ACCESS_TOKEN`.
```

## Troubleshooting

> ### How to check if runtime metrics are enabled?

Reviewer 2: ```suggestion
### Check if runtime metrics are enabled
```

> The SignalFx Instrumentation for .NET logs the profiling configuration at `INF` during the startup. 

Reviewer 2: ```suggestion
The SignalFx Instrumentation for .NET logs the profiling configuration using `INF` log messages during startup. 
```

> Verify that `runtime_metrics_enabled` key is set to `true` under `TRACER CONFIGURATION`.

Reviewer 2: ```suggestion
Verify that the `runtime_metrics_enabled` key is set to `true` under `TRACER CONFIGURATION`.
```

> ### How to check where are metrics exported?

Reviewer 2: ```suggestion
### Check where the metrics are being exported
```

> Check the value of the `metrics_agent_url` key under `TRACER CONFIGURATION`.

Reviewer 2: What does it look like if the export is done locally? (Not sure if it's possible)

Author: it looks like this `http://localhost:9943/v2/datapoint`

> ### How to verify collector setup?

Reviewer 2: ```suggestion
### Verify the Collector setup
```

> * Make sure you are running the [Splunk OpenTelemetry Collector](https://github.com/signalfx/splunk-otel-collector)

Reviewer 2: ```suggestion
* Make sure that the [Splunk OpenTelemetry Collector](https://github.com/signalfx/splunk-otel-collector) is running.
```

> * Make sure that an `signalfx` receiver and `signalfx` exporter are configured in the collector.

Reviewer 2: ```suggestion
* Make sure that a `signalfx` receiver and a `signalfx` exporter are configured in the Collector.
```

> Ensure that the `access_token` and `realm` fields are [correctly configured](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/signalfxexporter#metrics-configuration).

Reviewer 2: ```suggestion
* Make sure that the `access_token` and `realm` fields are [configured](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/signalfxexporter#metrics-configuration).
```

> Lastly, double check that the metrics pipeline is configured to use

Reviewer 2: ```suggestion
* Check that the metrics pipeline is configured to use
```
